File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/ImageUtils.java
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
 */

package org.openmrs.mobile.utilities;

import android.content.Context;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.widget.ImageView;

import androidx.core.content.ContextCompat;
import androidx.core.widget.ImageViewCompat;

import org.openmrs.mobile.R;
import org.openmrs.mobile.activities.patientdashboard.details.PatientPhotoActivity;

import java.io.ByteArrayOutputStream;

public final class ImageUtils {

    private ImageUtils() {

    }

    public static Bitmap decodeBitmapFromResource(Resources res, int resId, int reqWidth, int reqHeight) {
        // First decode with inJustDecodeBounds=true to check dimensions
        final BitmapFactory.Options options = new BitmapFactory.Options();
        options.inJustDecodeBounds = true;
        BitmapFactory.decodeResource(res, resId, options);

        // Calculate inSampleSize
        options.inSampleSize = calculateInSampleSize(options, reqWidth, reqHeight);

        // Decode bitmap with inSampleSize set
        options.inJustDecodeBounds = false;
        return BitmapFactory.decodeResource(res, resId, options);
    }

    public static int calculateInSampleSize(
            BitmapFactory.Options options, int reqWidth, int reqHeight) {
        // Raw height and width of image
        final int height = options.outHeight;
        final int width = options.outWidth;
        int inSampleSize = 1;

        if (height > reqHeight || width > reqWidth) {

            // Calculate ratios of height and width to requested height and width
            final int heightRatio = Math.round((float) height / (float) reqHeight);
            final int widthRatio = Math.round((float) width / (float) reqWidth);

            // Choose the smallest ratio as inSampleSize value, this will guarantee
            // a final image with both dimensions larger than or equal to the
            // requested height and width.
            inSampleSize = heightRatio < widthRatio ? heightRatio : widthRatio;
        }

        return inSampleSize;
    }

    public static Bitmap resizePhoto(Bitmap photo) {
        double HEIGHT = 500;
        double WIDTH = 500;
        int height = photo.getHeight();
        int width = photo.getWidth();
        double aspectRatio = Math.min(HEIGHT / height, WIDTH / width);

        if (0 < aspectRatio && aspectRatio < 1) {
            return Bitmap.createScaledBitmap(photo, (int) (aspectRatio * width), (int) (aspectRatio * height), true);
        }
        return photo;
    }

    public static Bitmap rotateImage(Bitmap source, float angle) {
        Matrix matrix = new Matrix();
        matrix.postRotate(angle);
        return Bitmap.createBitmap(source, 0, 0, source.getWidth(), source.getHeight(), matrix, true);
    }

    public static void showPatientPhoto(Context context, Bitmap photo, String patientName) {
        Intent intent = new Intent(context, PatientPhotoActivity.class);
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        photo.compress(Bitmap.CompressFormat.PNG, 0, byteArrayOutputStream);
        intent.putExtra("photo", byteArrayOutputStream.toByteArray());
        intent.putExtra("name", patientName);
        context.startActivity(intent);
    }

    public static void changeImageViewTint(Context context, ImageView imageView, int color) {
        ImageViewCompat.setImageTintList(imageView, ColorStateList.valueOf(ContextCompat.getColor(context, color)));

    }
}

File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/ImageUtils.kt
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
 */
package org.openmrs.mobile.utilities

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.ExifInterface
import com.openmrs.android_sdk.library.OpenmrsAndroid
import com.openmrs.android_sdk.utilities.ApplicationConstants.INTENT_KEY_NAME
import com.openmrs.android_sdk.utilities.ApplicationConstants.INTENT_KEY_PHOTO
import com.openmrs.android_sdk.utilities.ImageUtils
import org.openmrs.mobile.activities.patientdashboard.details.PatientPhotoActivity
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Date

object ImageUtils {

    @JvmStatic
    fun showPatientPhoto(context: Context, photo: Bitmap, patientName: String?) {
        val intent = Intent(context, PatientPhotoActivity::class.java)
        val byteArrayOutputStream = ByteArrayOutputStream()
        photo.compress(Bitmap.CompressFormat.PNG, 0, byteArrayOutputStream)
        intent.putExtra(INTENT_KEY_PHOTO, byteArrayOutputStream.toByteArray())
        intent.putExtra(INTENT_KEY_NAME, patientName)
        context.startActivity(intent)
    }

    fun createUniqueImageFileName(): String {
        val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss").format(Date())
        return timeStamp + "_" + ".jpg"
    }

    fun getResizedPortraitImage(imagePath: String): Bitmap {
        var portraitImg: Bitmap
        val options = BitmapFactory.Options().apply { inSampleSize = 4 }
        val photo = BitmapFactory.decodeFile(imagePath, options)
        val rotateAngle: Float
        try {
            val orientation = ExifInterface(imagePath).getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_UNDEFINED)
            rotateAngle = when (orientation) {
                ExifInterface.ORIENTATION_ROTATE_270 -> 270f
                ExifInterface.ORIENTATION_ROTATE_180 -> 180f
                ExifInterface.ORIENTATION_ROTATE_90 -> 90f
                else -> 0f
            }
            portraitImg = ImageUtils.rotateImage(photo, rotateAngle)
        } catch (e: IOException) {
            OpenmrsAndroid.getOpenMRSLogger().e(e.message)
            portraitImg = photo
        }
        return ImageUtils.resizePhoto(portraitImg)
    }
}
